lib/tools/adb-commands.js

```diff
@@ -262,7 +265,207 @@ methods.push = async function (localPath, remotePath) {
 };
 
 methods.pull = async function (remotePath, localPath) {
-  await this.exec(['pull', remotePath, localPath]);
+  await this.adbExec(['pull', remotePath, localPath]);
+};
+
+methods.processExists = async function (processName) {
+  try {
+    if (!this.isValidClass(processName)) {
+      log.errorAndThrow(`Invalid process name: ${processName}`);
+    }
+    let stdout = await this.shell("ps");
+    for (let line of stdout.split(/\r?\n/)) {
+      line = line.trim().split(/\s+/);
+      let pkgColumn = line[line.length - 1];
+      if (pkgColumn && pkgColumn.indexOf(processName) !== -1) {
+        return true;
+      }
+    }
+    return false;
+  } catch (e) {
+    log.errorAndThrow(`Error finding if process exists. Original error: ${e.message}`);
+  }
+};
+
+methods.forwardPort = async function (systemPort, devicePort) {
+  log.debug(`Forwarding system: ${systemPort} to device: ${devicePort}`);
+  await this.adbExec('forward', [`tcp:${systemPort}`, `tcp:${devicePort}`]);
+};
+
+methods.forwardAbstractPort = async function (systemPort, devicePort) {
+  log.debug(`Forwarding system: ${systemPort} to abstract device: ${devicePort}`);
+  await this.adbExec('forward', [`tcp:${systemPort}`, `localabstract:${devicePort}`]);
+};
+
+methods.ping = async function () {
+  try {
+    let stdout = await this.shell(["echo", "ping"]);
+    if (stdout.indexOf("ping") === 0) {
+      return true;
+    }
+    log.errorAndThrow(`ADB ping failed, returned ${stdout}`);
```

Comment:
since this will be caught in the catch here, we don't need to `errorAndThrow` it, we just need to throw it, it will be logged in the block below